### PR TITLE
Bug 1161249 - Add a link in treeherder  to perfherder graphs

### PR DESCRIPTION
**Pull request by MikeLing**
[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1009)

### CODE REVIEW

**Comment by MikeLing** [comment]
screen shot 2015-09-28 at 6 37 43 pm
I think we can retrieve datums by job id now :)

File: treeherder/webapp/api/performance_data.py [hunk 1]
```diff
@@ -83,20 +83,23 @@ def list(self, request, project):
 
         try:
             signature_hashes = request.QUERY_PARAMS.getlist("signatures")
+            job_id = request.QUERY_PARAMS.getlist("jobId")
```
Comment (wlach) [inline treeherder/webapp/api/performance_data.py]: I'd prefer the parameter `job_id` here.

File: ui/plugins/controller.js [hunk 2]
```diff
@@ -87,7 +87,7 @@ treeherder.controller('PluginCtrl', [
                     $scope.eta = $scope.job.get_current_eta();
                     $scope.eta_abs = Math.abs($scope.job.get_current_eta());
                     $scope.typical_eta = $scope.job.get_typical_eta();
-                    $scope.jobRevision = ThResultSetStore.getSelectedJob($scope.repoName).job.revision
+                    $scope.jobRevision = ThResultSetStore.getSelectedJob($scope.repoName).job.revision;
```
Comment (wlach) [inline ui/plugins/controller.js]: This is a good change but does not belong in this PR. Could you open up a new PR for this? I don't think we need to file a bug, since it's just a trivial change.

File: treeherder/webapp/api/performance_data.py [hunk 3]
```diff
@@ -83,20 +83,23 @@ def list(self, request, project):
...
+        if signature_hashes:
+            datums = PerformanceDatum.objects.filter(
+                repository=repository,
+                signature__signature_hash__in=signature_hashes).select_related(
                 'signature__signature_hash')
-
+        else:
+            datums = PerformanceDatum.objects.filter(
+                repository=repository,
+                job_id__in=job_id)
```
Comment (wlach) [inline treeherder/webapp/api/performance_data.py]: Does `job_id=job_id` not work?

Comment (MikeLing) [inline treeherder/webapp/api/performance_data.py]: yep, I think job_id=job_id can work as well. But job_id__in can work better ;)

File: ui/js/models/job_id.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                        } else {
+                            return $q.reject("No data been found for job id " +
+                                                  jobId + " in project " + projectName);
+                        }
+                    }
+                );
+            }
+        };
+    }
+]);
```
Comment (MikeLing) [inline ui/js/models/job_id.js]: Hi Will, could you give me some advice or hit about this? It seems like I fail to retrieve job id data from remote. @wlach

File: treeherder/webapp/api/performance_data.py (hunk 3 above)
Comment (wlach) [inline treeherder/webapp/api/performance_data.py]: Good point.

File: treeherder/webapp/api/performance_data.py [hunk 5]
```diff
@@ -84,14 +84,18 @@ def list(self, request, project):
 
         try:
             signature_hashes = request.query_params.getlist("signatures")
+            job_id = request.query_params.getlist("job_id")
```
Comment (wlach) [inline treeherder/webapp/api/performance_data.py]: This should be job_ids, since it's an array.

File: ui/js/models/job_id.js (hunk 4 above)
Comment (wlach) [inline ui/js/models/job_id.js]: Maybe try opening the developer console (https://developer.mozilla.org/en-US/docs/Tools/Web_Console) and see what URL is accessed? Then open that URL in a new window.

BTW, I don't think creating a new factory is the right approach here. I'd add a method for getting performance data by job id to the PhSeries factory in perf.js and include/import/use that.

File: ui/index.html [hunk 6]
```diff
@@ -106,6 +106,7 @@
         <script src="js/models/option_collection.js"></script>
         <script src="js/models/user.js"></script>
         <script src="js/models/error.js"></script>
+        <script src="js/perf.js"></script>
         <!-- Controllers -->
```
Comment (MikeLing) [inline ui/index.html]: I always got "Error: [$injector:unpr] Unknown provider: PhSeriesProvider <- PhSeries <- PluginCtrl" error, anything I miss to include or import to use the PhSeries? @wlach

Comment (wlach) [inline ui/index.html]: Yes, if you look at perf.js, we're adding PhSeries to the perf module (https://github.com/mozilla/treeherder/blob/master/ui/js/perf.js#L5).

It's probably time to split out PhSeries and move it to its own file in ui/js/models/ and add it to the treeherder module instead. As long as we import it in perf.html we should be fine, since the perf module depends on the treeherder module.

You might need to do a few other things as part of this process but I think that covers most of it. :)

File: ui/plugins/controller.js [hunk 7]
```diff
@@ -146,7 +152,11 @@ treeherder.controller('PluginCtrl', [
                     // the fourth result comes from the jobLogUrl artifact
                     // exclude the json log URLs
                     $scope.job_log_urls = _.reject(results[3], {name: 'mozlog_json'});
-
+                    $scope.perf_job_url = ('http://local.treeherder.mozilla.org/perf.html#/graphs?series=[' +
+                    $scope.repoName+ ',' + $scope.job.signature + ',1]&selected=[' +
+                    $scope.repoName + ',' + $scope.job.signature + ',' + $scope.job.result_set_id +
+                        ',' + $scope.job.id + ']');
+                    console.log($scope.job.signature)
```
Comment (MikeLing) [inline ui/plugins/controller.js]: Hi Will, I found the $scope.job.signature wasn't in the signature list which been retrieved by job_id endpoint, but they do share a same job_id(I mean the signature list been searched depend on job_id and the job_is is come from the 'job' we passed in). Is that possible? Can a job's signature miss by signature list which been searched by its job id?
@wlach

Comment (wlach) [inline ui/plugins/controller.js]: Right, a job signature is not the same thing as a performance signature (it's used to identify a job, as you might expect).

I made it possible last week to query performance data for a job (see bug 1214860). E.g.:

https://treeherder.mozilla.org/api/project/mozilla-inbound/performance/data/?job_id=15891075

You'll need to look up the signatures for all those jobs seperately, of course, maybe using a similar method as we do in the compare view:

https://github.com/mozilla/treeherder/blob/master/ui/js/perf.js#L448

File: ui/js/models/ph_series.js [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+            return {
+                platform: platform,
+                seriesList: seriesList,
+                testList: testList
+            };
+        });
+    };
+
+    var _getSeriesByJobId = function(projectName, jobId) {
```
Comment (MikeLing) [inline ui/js/models/ph_series.js]: we retrieve signature list which has same job id in here @jmaher

File: ui/plugins/controller.js [hunk 9]
```diff
@@ -76,11 +77,15 @@ treeherder.controller('PluginCtrl', [
                     job_id,
                     {timeout: selectJobPromise});
 
+                var jobIdPromise = PhSeries.getSeriesByJobId(
+                    $scope.repoName, job_id);
+
```
Comment (MikeLing) [inline ui/plugins/controller.js]: job id been used in here, and you can see the job id has been already been defined in the codebase which I only need to use it. @jmaher

File: ui/plugins/controller.js [hunk 10]
```diff
@@ -88,7 +93,11 @@ treeherder.controller('PluginCtrl', [
                     $scope.eta_abs = Math.abs($scope.job.get_current_eta());
                     $scope.typical_eta = $scope.job.get_typical_eta();
                     $scope.jobRevision = ThResultSetStore.getSelectedJob($scope.repoName).job.revision
-
+                    $scope.job_ids = results[4];
+                    console.log(results);
+                    _.forEach($scope.job_ids, function(job) {
+                        console.log(job[0].value);
+                    });
```
Comment (MikeLing) [inline ui/plugins/controller.js]: screen shot 2015-10-31 at 11 37 10 am
Hi Will, you can see the ts_paint and sessionrestore test missed in the picture which I think they shouldn't. And you can see I out put each value we got by job id:

> 1056, 244.77, 737.89, 1056, 1609.5, 2145, 798, 2145, 338, 798, 244.77

so their value haven't been found in the signature object list we retrieved by job id either.
@wlach

File: ui/plugins/controller.js [hunk 11]
```diff
@@ -141,6 +154,32 @@ treeherder.controller('PluginCtrl', [
...
+                                            // compare 14.10 with 14.12 and actually they point to the same job.
+                                            if (perfdata[0].value.toFixed(1) == Number(test.result).toFixed(1)) {
+                                                var data = {};
+                                                if (_.some($scope.perf_job_deatils, 'name', testName) === false) {
+                                                    data['name'] = testName;
+                                                    data['value'] = test.result;
+                                                    data['url'] = generateUrlForPerf($scope.repoName,
+                                                        perfdata[0].job_id, key, perfdata[0].result_set_id);
+                                                    $scope.perf_job_deatils.push(data);
+                                                }
```
Comment (MikeLing) [inline ui/plugins/controller.js]: screen shot 2015-10-31 at 11 47 03 am
you can see  tp5o_main_rss_paint and tp5o_modlistbytes_paint can't been displayed because I can't compare 'MB' with raw data. As I comment under the bug, I'm not sure if it's the right way to pick up signature we want by 'value'.

Comment (wlach) [inline ui/plugins/controller.js]: Yeah, don't worry about that for now. Just print out the raw value in bytes. We can write a formatter for this type of data later, when we encode it.

File: ui/plugins/controller.js (hunk 10 above)
Comment (wlach) [inline ui/plugins/controller.js]: What is the job id and project name you're seeing this on? It's hard for me to diagnose what's going on without that information.

Comment (MikeLing) [inline ui/plugins/controller.js]: screen shot 2015-11-01 at 9 24 54 am
So, I pick up another example. The project name is **mozilla-inbound** and the job id is **16628134**. The link of job id is https://treeherder.mozilla.org/api/project/mozilla-inbound/performance/data/?job_id=16628134.

Comment (wlach) [inline ui/plugins/controller.js]: So looking at the json I see at least signature `32206df28ab3b52e15d26f73178ee78e2af7e760` which corresponds to ts_paint. 

So if you're not seeing it in your output, there must be a bug in the way you're formatting things. Let me know if you need help tracking this down, but I suspect it's a pretty simple problem. :)

File: ui/js/models/ph_series.js [hunk 12]
```diff
@@ -0,0 +1,230 @@
+"use strict";
```
Comment (wlach) [inline ui/js/models/ph_series.js]: Oh one other comment: factoring out a file like this is probably a good idea, but you should do that in a distinct commit. Moving stuff to a file while making changes at the same time makes it difficult to see exactly what changed when reviewing. I'd put this back into perf.js for now. We can move the code in a followup commit.

File: ui/plugins/controller.js (hunk 10 above)
Comment (MikeLing) [inline ui/plugins/controller.js]: > So looking at the json I see at least signature 32206df28ab3b52e15d26f73178ee78e2af7e760 which corresponds to ts_paint. 

yes, we can get corresponds signature to ts_paint, but can't distinguish it with other signature from a bunch of data. Because the job value of  32206df28ab3b52e15d26f73178ee78e2af7e760 is **1132.00** but the value we want for ts_paint is 1144.70. 

However, let me sort the difficult I meet, maybe it can help us set it ASAP :)
The fist step I do is get a bunch of signature object by job_id.  Then I unpack the job_detail object and pick up the job we want to display(I assume we only need to display the 'talos result' job). So we can find the value for each job(eg. 677.65 for a11yr_paint, 2199.78 for sessionrestore)

At last, we can compare each job value in signature object with we found in job detail. If we got two *same* value, then we can assume they are same job and could be used to generate a link to perherder.

**So, some difficulties I have:**
**1) we can't convert some value like 'MB' to pure raw data (pass)**
**2) I don't know the correct way to pick up the right signature from signature list. Although, we can get the correct signature list. Otherwise, we need to display all the signature we got by job id.**
**ps**: I always found pick up data by compare job value is somehow inaccuracy. maybe it's a **wrong** way to deal with this issue. and maybe we should just looking up each signature and display them like: 

suite: url to perf 
rather than 
testName: value(url to perf)

But there are too many things we need to display which isn't right also, I think

>So if you're not seeing it in your output, there must be a bug in the way you're formatting things. Let me know if you need help tracking this down, but I suspect it's a pretty simple problem. :)

Yes, I just feel like jump into a 'cul-de-sac' and need to some hits to get out. always wonder if the method I use is right and can't figure out the right way to do it either. Great appreciate if you can help me out! Thank you!

File: ui/js/models/ph_series.js (hunk 12 above)
Comment (MikeLing) [inline ui/js/models/ph_series.js]: Ok, I will commit another PR for this change. Then I can pull and rebase it after you merge it :)

File: ui/plugins/controller.js (hunk 10 above)
Comment (wlach) [inline ui/plugins/controller.js]: So I'm still a bit confused about your confusion. Let me try to break down how I would go about solving this, maybe it will help:

So if you look up this:

https://treeherder.mozilla.org/api/project/mozilla-inbound/performance/signatures/?signature=32206df28ab3b52e15d26f73178ee78e2af7e760

You get this response:

```
{
    "32206df28ab3b52e15d26f73178ee78e2af7e760": {
        "suite": "ts_paint",
        "subtest_signatures": [
            "726a741fd1dd1e9fb0b85bc9733658f6eabd07df"
        ],
        "option_collection_hash": "102210fe594ee9b33d82058545b1ed14f4c8206e",
        "machine_platform": "linux64"
    }
}
```

You see that the signature id is right there in the dictionary. You can just look up the properties based on the signature hash and then create the name of the series from that. To get all the signatures, just do a bigger query:

https://treeherder.mozilla.org/api/project/mozilla-inbound/performance/signatures/?signature=<signature hash1>&signature=<signature hash 2>...

Then write a routine to unpack the values and associate them with particular series.

**Comment by MikeLing** [comment]
screen shot 2015-11-03 at 7 31 50 pm
screen shot 2015-11-03 at 7 32 07 pm
ok, we finally make it better than last time :) @wlach @jmaher

**Comment by jmaher** [comment]
nice!

File: ui/js/models/ph_series.js [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                seriesList: seriesList,
+                testList: testList
+            };
+        });
+    };
+
+    var _getSeriesByJobId = function(projectName, jobId) {
+        var JobIdURL = thServiceDomain + '/api/project/' + projectName +
+            '/performance/data/?job_id=' + jobId;
+        return $http.get(JobIdURL).then(function(response) {
```
Comment (wlach) [inline ui/js/models/ph_series.js]: It shouldn't be necessary to create a temporary variable called jobIdURL here -- just pass it directly to $http.get.

Also the first letter of a variable like this should be uncapitalized. I.e. `JobId` -> `jobId`

File: ui/plugins/controller.js [hunk 14]
```diff
@@ -141,6 +152,59 @@ treeherder.controller('PluginCtrl', [
...
+                            $q.all(_.chunk(signatureList, 20).map(function(signatureChunk) {
+                                var url = '/performance/signatures/?signature=' + signatureChunk.pop();
+                                signatureChunk.forEach(function(signature) {
+                                    url = url + '&signature=' + signature;
+                                });
+                                return $http.get(thUrl.getProjectUrl(url, $scope.repoName)).then(
+                                    function(response) {
+                                        // In general, we only need to show the summary signature. But
+                                        // some other test like 'main_rss' of tp5o been required also.
+                                        // For now, I only know 'Private Bytes', 'XRes' and 'Main_RSS'
```
Comment (wlach) [inline ui/plugins/controller.js]: We should probably do the same thing we do for compare view here: get a giant list of all the signatures, and then filter out those subtests which have a summary associated with them:

https://github.com/mozilla/treeherder/blob/master/ui/js/perf.js#L171

Any hardcoded list like this is bound to be incomplete and out of date.

File: ui/plugins/controller.js [hunk 15]
```diff
@@ -141,6 +152,56 @@ treeherder.controller('PluginCtrl', [
...
+                                            serie['signature'] = key;
+                                            series.push(serie);
+                                        })
+                                    }
+                                );
+                            })).then(function(){
+                                var allSubtestSignatures = _.flatten(_.map(series, function(serie) {
+                                    return serie.subtest_signatures ? serie.subtest_signatures : [];
+                                }));
+                                _.forEach(series, function(serie) {
```
Comment (wlach) [inline ui/plugins/controller.js]: serie -> series

File: ui/plugins/controller.js [hunk 16]
```diff
@@ -141,6 +152,56 @@ treeherder.controller('PluginCtrl', [
...
+                                    if (!serie.subtest_signatures) {
+                                        if (_.contains(allSubtestSignatures, serie.signature)) {
+                                            return;
+                                        }
+                                    }
+                                    if (serie.test) {
+                                        if (serie.suite == 'tp5o' && !_.contains(['Private Bytes',
+                                                'XRes', 'Main_RSS', 'shutdown'], serie.test)) {
+                                            return;
```
Comment (wlach) [inline ui/plugins/controller.js]: I don't understand why this is necessary?

File: ui/plugins/talos/main.html [hunk 17]
```diff
@@ -10,7 +10,14 @@
       </ul>
     </li>
   </ul>
-
+  <ul ng-show="perf_job_detail">
+    <li>Perfherder:
+      <ul ng-repeat="line in perf_job_detail | toArray | orderBy: '$key'">
```
Comment (wlach) [inline ui/plugins/talos/main.html]: orderBy: '$key' should be orderBy: 'suite'. $key isn't defined for the perfherder array.

File: ui/plugins/controller.js [hunk 18]
```diff
@@ -88,6 +94,7 @@ treeherder.controller('PluginCtrl', [
                     $scope.eta_abs = Math.abs($scope.job.get_current_eta());
                     $scope.typical_eta = $scope.job.get_typical_eta();
                     $scope.jobRevision = ThResultSetStore.getSelectedJob($scope.repoName).job.revision;
+                    $scope.job_ids = results[4];
```
Comment (wlach) [inline ui/plugins/controller.js]: `job_ids` -> `jobIds`

File: ui/plugins/controller.js [hunk 19]
```diff
@@ -123,7 +130,11 @@ treeherder.controller('PluginCtrl', [
                     $scope.jobSearchSignature = $scope.job.signature;
                     $scope.jobSearchStrHref = getJobSearchStrHref($scope.jobSearchStr);
                     $scope.jobSearchSignatureHref = getJobSearchStrHref($scope.job.signature);
-
+                    var generateUrlForPerf = function(projectName, jobId, signature, resultSetId) {
```
Comment (wlach) [inline ui/plugins/controller.js]: This function is only called once, so I don't think it's necessary. Just put the code inline where it's used.

File: ui/plugins/controller.js [hunk 20]
```diff
@@ -141,6 +152,46 @@ treeherder.controller('PluginCtrl', [
                           }
                           return result;
                         }, []);
+                        if ($scope.job_ids !== null) {
+                            var signatureList = _.keys($scope.job_ids);
+                            var seriesList = [];
+                            $scope.perf_job_detail = [];
```
Comment (wlach) [inline ui/plugins/controller.js]: `perf_job_detail` -> `perfJobDetail`.

In general always use camel case when defining a variable in JavaScript.

**Comment by wlach** [comment]
Just tested this, it works great! Thanks for doing the work here. Just a few more nits and we can land.

**Comment by MikeLing** [comment]
Oh, I forget to tell you I had address those nits. @wlach ;)

File: ui/plugins/controller.js [hunk 21]
```diff
@@ -141,6 +147,48 @@ treeherder.controller('PluginCtrl', [
...
+                                    detail['url'] = thServiceDomain + '/perf.html#/graphs?series=[' +
+                                        $scope.repoName+ ',' + series.signature + ',1]&selected=[' +
+                                        $scope.repoName + ',' +  series.signature + ',' +
+                                        $scope.job['result_set_id'] + ',' + $scope.job['id'] + ']';
+                                    detail['value'] = $scope.jobIds[series.signature][0].value;
+                                    detail['suite'] = series.suite;
+                                    if (series.hasOwnProperty('test')) {
+                                        detail['suite'] = series.suite + '_' +
+                                            series.test.toLowerCase();
+                                    }
```
Comment (wlach) [inline ui/plugins/controller.js]: Hey, I just tested this on stage where I just landed a bug to remove the summary signatures for suites with only one test in them, and the results no longer look that good in that case:

screen shot 2015-11-06 at 12 57 18 pm

I think we should remove the subtest details for that case. Something like this should do the trick:

````
detail['value'] = $scope.jobIds[series.signature][0].value;
detail['title'] = series.suite;
if (series.hasOwnProperty('test') && series.test.toLowerCase() !== series.suite) {
    detail['title'] += '_' + series.test.toLowerCase();
}
```

Note the change from suite -> title, which I believe is more accurate (you'll need to change the .html accordingly also, of course). I think with that change we should be good, sorry to make you have to do a bit more work. :)

Comment (wlach) [inline ui/plugins/controller.js]: To test this, just change your local.conf.js to point to 'stage' instead of 'production'.

File: ui/plugins/talos/main.html [hunk 22]
```diff
@@ -10,7 +10,14 @@
       </ul>
     </li>
   </ul>
-
+  <ul ng-show="perfJobDetail">
+    <li>Perfherder:
+      <ul ng-repeat="line in perfJobDetail | toArray | orderBy: 'suite'">
```
Comment (wlach) [inline ui/plugins/talos/main.html]: You forgot to change 'suite' to 'title' here, when making the corresponding change in controller.js, so the results are no longer sorted.

I'll fix this up for you and push (since we're so close) but try and bear this sort of thing in mind for the future. I can't always tell you *everything* that needs to be done in a review comment, I partly depend on you to test stuff and make sure it all looks good before landing. :)

**Comment by wlach** [comment]
Merged in 3f96dc6d7bd26bfd1a455999ea3d0869f676252c
